From 74b17ab5cf143fa7d6136d69301c23c305edb901 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Wed, 29 Oct 2014 15:47:35 +0800
Subject: [PATCH 0307/5242] MLK-9760: ASoC: fsl_esai: fix NULL pointer issue
 in reset handler

commit  447e7ea5ebe5d22eeab96e34025f13ba9add045a from
https://source.codeaurora.org/external/imx/linux-imx.git

When test with case arecord -Dhw:0,1 | aplay -Dhw:0,0, xrun happened,
the reset handler will be called, but for BE(backend) stream, the
substream->ops is null.
This patch is to fix this null pointer issue.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
(cherry picked from commit 4db112a8cd3caf5a553afea88cf7fe8d9781f459)
(cherry picked from commit cf060b0f735d0fadadda2dc03d0b38e96ca7ffe2)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_esai.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/sound/soc/fsl/fsl_esai.c b/sound/soc/fsl/fsl_esai.c
index 3f86588..b48f25b 100644
--- a/sound/soc/fsl/fsl_esai.c
+++ b/sound/soc/fsl/fsl_esai.c
@@ -813,7 +813,8 @@ static int stop_lock_stream(struct snd_pcm_substream *substream)
 {
 	if (substream) {
 		snd_pcm_stream_lock_irq(substream);
-		if (substream->runtime->status->state == SNDRV_PCM_STATE_RUNNING)
+		if (substream->runtime->status->state == SNDRV_PCM_STATE_RUNNING
+			&& substream->ops)
 			substream->ops->trigger(substream, SNDRV_PCM_TRIGGER_STOP);
 	}
 	return 0;
@@ -822,7 +823,8 @@ static int stop_lock_stream(struct snd_pcm_substream *substream)
 static int start_unlock_stream(struct snd_pcm_substream *substream)
 {
 	if (substream) {
-		if (substream->runtime->status->state == SNDRV_PCM_STATE_RUNNING)
+		if (substream->runtime->status->state == SNDRV_PCM_STATE_RUNNING
+			&& substream->ops)
 			substream->ops->trigger(substream, SNDRV_PCM_TRIGGER_START);
 		snd_pcm_stream_unlock_irq(substream);
 	}
-- 
1.7.9.5

